include/permanent.hpp

```diff
@@ -322,34 +322,253 @@ std::complex<double> permanent_quad(std::vector<std::complex<double>> &mat) {
+ * and returns sensible values for empty and non-even matrices.
  *
  *
  * @param mat vector representing the flattened matrix
  * @return the permanent
  */
-double permanent_fsum(std::vector<double> &mat) {
+double permanent_fsum(std::vector<double> &mat) 
+{
     std::vector<double> matq(mat.begin(), mat.end());
     double perm = perm_fsum(matq);
     return static_cast<double>(perm);
 }
 
+
+/**
+ * Returns the permanent of a matrix (nthreads=1)
+ *
+ * \rst
+ *
+ * Returns the permanent of a matrix using the BBFG algorithm.
+ * This algorithm was given by Glynn (2010) with the time-complexity 
+ * of \f$O(n 2^n)\f$ using Gray code ordering.
+ *
+ * \endrst
+ *
+ *
+ * @param mat  a flattened vector of size \f$n^2\f$, representing an
+ *      \f$n\times n\f$ row-ordered symmetric matrix.
+ * @return permanent of the input matrix
+ */
+template <typename T>
+inline T perm_BBFG_serial(std::vector<T> &mat) 
+{
+    int n = std::sqrt(static_cast<double>(mat.size()));
+    llint x = static_cast<llint>(pow(2,n - 1));
+    std::vector<T> col_sum(n, static_cast<T>(0));
+
+    // init col_sum 
+    for (int i=0; i < n; ++i) {
+        for (int j=0; j < n; ++j) {
+            col_sum[i] += mat[j*n + i];
+        }
+    }
+
+    constexpr T p1 = static_cast<T>(1.0);
+    constexpr T p2 = static_cast<T>(2.0);
+    constexpr T n2 = static_cast<T>(-2.0);
+
+    T mulcolsum, coeff;
+    T total = static_cast<T>(0);
+    llint og=0, ng, gd;
+    int sgn=1, gdi;
+
+    // over all 2^{n-1} permutations of delta
+    for (llint k=1; k < x+1; ++k) {
+        mulcolsum = std::accumulate(
+                        col_sum.begin(), 
+                        col_sum.end(), 
+                        p1, 
+                        std::multiplies<T>());
+        total += sgn > 0 ? mulcolsum : -mulcolsum;
+        
+        // updating gray order
+        ng = igray(k);
+        gd = og ^ ng;
+
+        coeff = og > ng ? p2 : n2;
+        gdi = left_most_set_bit(gd);
+        for (int j=0; j < n; ++j) {
+            col_sum[j] += coeff * mat[gdi * n + j];
+        }
+
+        sgn = -sgn;
+        og = ng;
+    }
+
+    // TODO: overflow handling
+
+    return total / static_cast<T>(x);
+}
+
+/**
+ * Returns the permanent of a matrix (nthreads=1) (2nd version)
+ *
+ * \rst
+ *
+ * Returns the permanent of a matrix using the BBFG algorithm.
+ * This algorithm was given by Glynn (2010) with the time-complexity 
+ * of \f$O(n 2^n)\f$ using Gray code ordering.
+ *
+ * \endrst
+ *
+ *
+ * @param mat  a flattened vector of size \f$n^2\f$, representing an
+ *      \f$n\times n\f$ row-ordered symmetric matrix.
+ * @return permanent of the input matrix 
+ */
+template <typename T>
+inline T perm_BBFG_serial2(std::vector<T> &mat)
+{
+    int n = std::sqrt(static_cast<double>(mat.size()));
+    int sgn=1, k=0;
+
+    constexpr T p2 = static_cast<T>(2.0);
+    constexpr T p1 = static_cast<T>(1.0);
+
+    int *gray_list = new int[n];
+    T *coeffs = new T[n];
```

Comment:
Any thoughts on how to improve memory safety here?

Reply:
As far as I know, there are two best practices to solve mem safety here: using 1) smart pointers  2) std containers. Here, I think the best approach would be using std containers instead of the former, e.g. `unique_ptr<T[]>` or `boost::shared_array`. Among std containers, I also think of `std::vector` as the size of `gray_list` and `coeffs` are fixed within the program but not a constant expression. What do you think?  Thank you for bringing it up here!

Reply:
In the long run I think `unique_ptr`, or just allocating a `vector` would be good here. Matching the existing implementations though, it should be fine for now. This may be something we can tackle in a later PR.